shared/ssz/hash_cache.go

```diff
@@ -0,0 +1,279 @@
+		Help: "The number of hashes in the block cache",
+	})
+)
+
+// markleRoot specifies the hash of data in a struct
+type root struct {
+	Hash       common.Hash
+	MarkleRoot []byte
+}
+
+// hashKeyFn takes the  representation as the key for a hashInfo.
+func hashKeyFn(obj interface{}) (string, error) {
+	mRoot, ok := obj.(*root)
+	if !ok {
+		return "", ErrNotMarakleRoot
+	}
+
+	return mRoot.Hash.Hex(), nil
+}
+
+// newHashCache creates a new hash cache for storing/accessing root hashes from
+// memory.
+func newHashCache() *hashCacheS {
+	return &hashCacheS{
+		hashCache: cache.NewFIFO(hashKeyFn),
+	}
+}
+
+// hashCacheS struct with one queue for looking up by hash.
+type hashCacheS struct {
+	hashCache *cache.FIFO
+}
+
+// RootByHash fetches Root by its hash. Returns true with a
+// reference to the root if exists. Otherwise returns false, nil.
+func (b *hashCacheS) RootByHash(h common.Hash) (bool, *root, error) {
+
+	obj, exists, err := b.hashCache.GetByKey(h.Hex())
+	if err != nil {
+		return false, nil, err
+	}
+
+	if exists {
+		hashCacheHit.Inc()
+	} else {
+		hashCacheMiss.Inc()
+		return false, nil, nil
+	}
+
+	hInfo, ok := obj.(*root)
+	if !ok {
+		return false, nil, ErrNotMarakleRoot
+	}
+
+	return true, hInfo, nil
+}
+
+// AddRetrieveTrieRoot adds a trie root to the cache. This method also trims the
+// least recently added root info if the cache size has reached the max cache
+// size limit.
+func (b *hashCacheS) AddRetrieveTrieRoot(val interface{}) ([32]byte, error) {
+	if val == nil {
+		return [32]byte{}, newHashError("untyped nil is not supported", nil)
+	}
+	rval := reflect.ValueOf(val)
+
+	startTime := time.Now().UnixNano()
+	hs, err := hashedEncoding(rval)
+	if err != nil {
+		return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
+	}
+	exists, fetchedInfo, err := b.RootByHash(bytesutil.ToBytes32(hs))
+	if err != nil {
+		return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
+	}
+	fmt.Printf("encoding time: %v \n", time.Now().UnixNano()-startTime)
+	var paddedOutput [32]byte
+	if exists {
+		paddedOutput = bytesutil.ToBytes32(fetchedInfo.MarkleRoot)
+	} else {
+		sszUtils, err := cachedSSZUtils(rval.Type())
+		if err != nil {
+			return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
+		}
+		startTime = time.Now().UnixNano()
+		output, err := sszUtils.hasher(rval)
+		fmt.Printf("hashing time: %v \n", time.Now().UnixNano()-startTime)
+		if err != nil {
+			return [32]byte{}, newHashError(fmt.Sprint(err), rval.Type())
+		}
+		// Right-pad with 0 to make 32 bytes long, if necessary
+		paddedOutput = bytesutil.ToBytes32(output)
+		b.AddRoot(bytesutil.ToBytes32(hs), paddedOutput[:])
+	}
+
+	return paddedOutput, nil
+}
+
+// AddRetriveMarkleRoot adds a mrakle object to the cache. This method also trims the
+// least recently added root info if the cache size has reached the max cache
+// size limit.
+func (b *hashCacheS) AddRetriveMarkleRoot(byteSlice [][]byte) ([]byte, error) {
```

Comment:
The naming is again confusing, and it indicates its doing the same exact thing as the above function

Reply:
fixed